src/hotspot/share/runtime/os.cpp

```diff
@@ -928,14 +928,55 @@ bool os::print_function_and_library_name(outputStream* st,
   return have_function_name || have_library_name;
 }
 
-ATTRIBUTE_NO_ASAN static void print_hex_readable_pointer(outputStream* st, address p,
-                                                         int unitsize) {
+ATTRIBUTE_NO_ASAN static bool read_safely_from(intptr_t* p, intptr_t* result) {
+  const intptr_t errval = 0x1717;
+  intptr_t i = SafeFetchN(p, errval);
+  if (i == errval) {
+    i = SafeFetchN(p, ~errval);
+    if (i == ~errval) {
+      return false;
+    }
+  }
+  (*result) = i;
+  return true;
+}
+
+static bool print_hex_location(outputStream* st, address p, int unitsize) {
+  intptr_t i = 0;
+  const intptr_t errval = 0x1717;
+  address pa = align_down(p, sizeof(intptr_t));
+  if (!read_safely_from((intptr_t*)pa, &i)) {
+    return false;
+  }
+#ifndef _LP64
+  // Special handling for printing qwords on 32-bit platforms
+  if (unitsize == 8) {
+    intptr_t i2 = 0;
+    assert(sizeof(intptr_t) == 4, "Sanity");
+    if (!read_safely_from((intptr_t*)pa + 1, &i2)) {
+      return false;
+    }
+    uint64_t value = (((uint64_t)i2) << 32) | i;
+    st->print("%016" FORMAT64_MODIFIER "x", value);
```

Comment:
This implies endianness, it is not? We handle endianness for 64-bit version below.

Reply:
Sure, for completeness; though I don't think we have a 32-bit big-endian platform.